### PR TITLE
Change Density test to avoid pulling image be the main factor

### PR DESCRIPTION
**Pull request by wojtek-t**
cc @fgrzadkowski

### CODE REVIEW

**Comment by zmerlynn** [comment]
If you're going to flip to pause, flip to the gcr.io image of it in the process. (On phone, but it's something like gcr.io/google-containers/pause cc @jlowdermilk)

I'm also not enamored with the 100 -> 30 swap, because @rrati was definitely hitting issues at 100

**Comment by wojtek-t** [comment]
Regarding 100 -> 30 swap, I did it consciously, because in v1.0 we are NOT targeting 100 pods per node. What we would like to support is 30 pods per node, thus I think we shouldn't try to optimize for this case now. (But I agree we can get back to it after v1.0).

**Comment by wojtek-t** [comment]
As for flipping the image, kubernetes/pause:go is the image that is used by Pod Infra Container. Thus it is already on the machine - that's why it will never be pulled for the "user container" itself.

**Comment by rrati** [comment]
I don't think we should drop the pod count here.  The density test I wrote can easily be modified to be multiple tests that scales up as desired, but we shouldn't drop from 100 imo.  100 pods/node may not be a target for 1.0, but that degree of density has shaken out a number of 1.0 blocking issues.  I don't feel 1.0 will be served better by dropping the count in this test.

iirc, 1.0 is targeting 30-50 pods/node.  This test should become multiple at various density levels imo.  30, 50, 100, etc.  That's easy enough to do with a for loop outside the It condition.

**Comment by wojtek-t** [comment]
I think that running 100 pods per node is exposing many more issues than those we currently care for v1.0. And if all of them will be a result of running this Density test, it would be difficult to say whether it's a problem also at the level of 30 pods or does it appear only with say more than 80 pods per node.

Basically, what I would like to do is to have a "regression test" that will be part of e2e suite.
And I think that "should allow starting 30 pods per node" with some expectations added to it would be a good starting point for it.

**Comment by wojtek-t** [comment]
Maybe a good solution for it would be:
- leave a 100 pod per node test as skipped by default
- add a 30 pod per node test as a test that is run as part of e2e suite.
What do you think?

**Comment by zmerlynn** [comment]
Density isn't run as part of default e2e right now anyways. Would this
change allow a really low time budget density test? If so, I'm supportive
of 30-in-e2e.. But the 100.. Do we need parametrization?

**Comment by wojtek-t** [comment]
Yes - I know currently it's not running. And the whole e2e test still takes too long to be part of e2e suite. However, the ""should allow starting 30 pods per node" scenario (after the image change) is taking less than a minute (I ran it 3 times and all of them were running 40-50 seconds) so I think that is reasonable. The 100 pods per node can run at least 4 minutes.

**Comment by wojtek-t** [comment]
To be clear - I'm happy to change all of it in this PR, but I'm waiting for some confirmation that you guys agree with that.

**Comment by zmerlynn** [comment]
At less than a minute, I have no problem with it just being an e2e (which
requires some other gymnastics to get it back out of the skip list :/)

**Comment by wojtek-t** [comment]
Thanks - I will try to change it.

**Comment by rrati** [comment]
I'm fine with 2 separate tests.  I think keeping the 100 pod/node even if it's not run as part of e2e is important.  We use it a lot.

We should be able to enable the density suite and exclude the 100 pod/node test during normal e2e runs.

**Comment by wojtek-t** [comment]
Thanks @rrati - SGTM. I will work on it and let you know once I update this PR.

**Comment by wojtek-t** [comment]
I've updated the PR - PTAL

also I'm running e2e tests - the new PerfTest already passed:
• [SLOW TEST:43.356 seconds]
PerfTest
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:209
  should allow starting 30 pods per node

**Comment by rrati** [comment]
Is there a need to create another test suite?  This is an awful lot of duplicated code.  Why not just add another It clause, make the suite non-pending, and exclude running the 100 pod/node test at e2e invocation?

**Comment by wojtek-t** [comment]
This test isn't pending - it's just excluded as the whole when running e2e suite. I agree it's not beautiful, so if you know how to do it better, can you please be more precise?

**Comment by wojtek-t** [comment]
Any comments/suggestions?

**Comment by rrati** [comment]
My thought was something like:

for cnt := range []int{30, 50 100} {
  It("...%d pods per node", cnt) {
    ...
  }
}

Not sure if ginkgo allows that type of definition, but other test frameworks have in other languages.  The above makes it easy to add another level of density to the test since the only difference between the tests is the number of pods/node.  The It clauses will be unique so that can still be filtered on.

**Comment by wojtek-t** [comment]
I agree that would be fine as long as the whole test is disabled. But what I would like to achieve is to enable the 30-pod-per-node test in our e2e runs. And that will not run because the whole Density test is disabled:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/driver.go#L73

That's why I created the separate test. IIUC, this will not work with what you're suggesting.

**Comment by rrati** [comment]
We could rework that exclude logic in the driver to call out the 100 pod/node test explicitly.  Are there other tests in the density suite we would not want to run?

Would it make sense to move to an explicit blacklist of tests we want to avoid in an e2e run?  It seems like the default behavior should be to run all e2e tests so it makes sense that we should require some effort to prevent a test from running (ie changing the blacklist)

**Comment by wojtek-t** [comment]
OK - I've refactored the test to avoid code duplication and only 30-pod-per node is enabled in e2e test suite.
PTAL

**Comment by rrati** [comment]
LGTM

**Comment by wojtek-t** [comment]
Zach - since you already commented on it, I'm assigning it to you. Feel free to reassign if you think someone else should also take a look into it.

File: test/e2e/density.go [hunk 1]
```diff
@@ -206,12 +207,18 @@ var _ = Describe("Density", func() {
...
 	})
 
-	It("should allow starting 100 pods per node", func() {
-		RCName = "my-hostname-density100-" + string(util.NewUUID())
-		RunRC(c, RCName, ns, "dockerfile/nginx", 100*minionCount)
-	})
+	for _, count := range []int{30, 50, 100} {
+		name := fmt.Sprintf("should allow starting %d pods per node", count)
+		if count > 30 {
+			name = "[Skipped] " + name
```
Comment (fgrzadkowski) [inline test/e2e/density.go]: Can you please add a comment that tests with "Skipped" in name will be skipped when running e2e without --ginkgo.focus flag? For me it's not obvious why you add this prefix.

**Comment by fgrzadkowski** [comment]
LGTM

File: test/e2e/density.go (hunk 1 above)
Comment (wojtek-t) [inline test/e2e/density.go]: Done.

**Pull request merged**